### PR TITLE
Add Graph Docker container to external services

### PR DESCRIPTION
**Pull request by TimDiekmann**
## 🌟 What is the purpose of this PR?

In order to run the Graph API as external service, it should be added to the external services. This enables easy usage of the Graph API without the need of having Rust installed or complicated dependencies between containers.

## 🔗 Related links

- [Asana task](https://app.asana.com/0/1201095311341924/1202771673621395) _(internal)_

## 🚫 Blocked by

- #926 

## 🔍 What does this change?

- Add the Docker file to `packages/hash/external-services`
- Run REST API tests on external service Graph (which avoids building/running the Dockerfile manually)
- Add healty-condition to postgres external service to only run the Graph when the DB is ready
- Drive-by: Use a volume for the postgres external service, so it does not deescalate rights when executing (SemGrep will still complain as it's not completely forbidden)
- Drive-by: Removed unused TS file and configurations

## 📜 Does this require a change to the docs?

No

## ⚠️ Known issues

The Docker container is built on the external services. In a future PR the Dockerimage should be built, which is then pulled from the CI for running the tests on `packages/hash` (for timings it does not make a difference, if building happens in the same Job).

## 🐾 Next steps

See above

## 🛡 What tests cover this?

- The REST endpoint tests using the Dockerfile

## ❓ How to test this?

Run the external services and ensure the Graph API is working

### CODE REVIEW

File: packages/graph/deployment/graph/Dockerfile
Comment (thehabbos007) [inline packages/graph/deployment/graph/Dockerfile:26]: nit and non blocking:
I think `/opt/` would be more appropriate, but I don't feel strongly, just a little shorter than this :)

File: packages/graph/deployment/graph/migrate_and_run.sh
Comment (thehabbos007) [inline packages/graph/deployment/graph/migrate_and_run.sh:13]: nit/discussion
We also define this URI in the root `.env`. Don't know if we should share these somehow, the only difference is that the URI in the root splits ups the hostname/ip and the port, which makes them incompatible without interpolation, unfortunately

File: packages/hash/external-services/docker-compose.yml
Comment (thehabbos007) [inline packages/hash/external-services/docker-compose.yml:51]: I think this volume mount was done on purpose in external services, but I'm not sure what for. Just flagging we should check if there was a good reason for this to be mounted beforehand

File: packages/hash/external-services/docker-compose.yml
Comment (thehabbos007) [inline packages/hash/external-services/docker-compose.yml:55]: This is so cool, thanks for finding this!

File: packages/hash/external-services/docker-compose.yml
Comment (thehabbos007) [inline packages/hash/external-services/docker-compose.yml:38]: I think we also want `init: true` on postgress. I think not everyone uses `detach`ed mode for the compose, so if having init makes ctrl+c possible with alpine, we should add that

File: packages/hash/external-services/docker-compose.yml
Comment (thehabbos007) [inline packages/hash/external-services/docker-compose.yml:56]: I think this healthcheck runs indefinitely, though, so a 1s interval seems a little too often if it checks forever. Perhaps 5s or 10s is better? Just means we will wait 5s or 10s on initial start before probing (might be fine if it takes a couple retries to suceed, though)

**Review by thehabbos007 — CHANGES_REQUESTED** [review]
Thanks Tim! Left some comments :)

File: packages/graph/.dockerignore
Comment (TimDiekmann) [reply packages/graph/.dockerignore]: Good point: 37ce55ec9d5b4c2bd6a21044ce73fc08455105f2

File: packages/graph/Makefile.toml
Comment (TimDiekmann) [reply packages/graph/Makefile.toml:12]: It's required for the `docker-compose.yml` files. `ROOT/.env.local` is used for external services, `packages/graph/.env.local` is used for the Graph one

File: packages/graph/deployment/graph/Dockerfile
Comment (TimDiekmann) [reply packages/graph/deployment/graph/Dockerfile:26]: I normally use `/opt` for full programs, i.e. binaries which require more files than only the binary itself. As both files are self-contained I put this in `/usr/local`. I don't really mind 🙂

File: packages/graph/deployment/graph/migrate_and_run.sh
Comment (TimDiekmann) [reply packages/graph/deployment/graph/migrate_and_run.sh]: Good point, used the environment variable instead: ce5381d19a496bfc4af09604ed074932b9f44835

File: packages/hash/external-services/docker-compose.yml
Comment (TimDiekmann) [reply packages/hash/external-services/docker-compose.yml:56]: The current check will run the check once per second, each second could take 1 second. After 5 retries, it will fail. Optionally, we could also set a start period, but I don't think it's needed. This way at least it tries to complete earlier because I don't really expect it to take several seconds to start. However, the number of retries could be raised, what do you think?

File: packages/hash/external-services/docker-compose.yml
Comment (TimDiekmann) [reply packages/hash/external-services/docker-compose.yml:51]: @benwerner01 do you know more about this?

File: packages/graph/deployment/graph/Dockerfile
Comment (TimDiekmann) [reply packages/graph/deployment/graph/Dockerfile:29]: We expect to move to a `FROM scratch` image later on, so I'm not going to fix this.

File: packages/hash/external-services/docker-compose.yml
Comment (TimDiekmann) [reply packages/hash/external-services/docker-compose.yml:38]: Somehow, for `postgres` this is not needed. It will quit without issues (Don't ask why 😄).

File: packages/hash/external-services/docker-compose.yml
Comment (benwerner01) [reply packages/hash/external-services/docker-compose.yml:51]: I'm not sure either

Comment (TimDiekmann) [reply packages/hash/external-services/docker-compose.yml:51]: @kachkaev do you know why this is mounting a directory rather than using a volume? The current way it's escalating root privileges and writes to `./var/external-services/...` with root rights.

File: packages/hash/external-services/docker-compose.yml
Comment (TimDiekmann) [reply packages/hash/external-services/docker-compose.yml:73]: How do we want to be explicit here? (This is one of the reasons why I really dislike global `.env` files and pulling everything from there) Currently, we don't specify `.env.local` at all when invoking `docker-compose`, so it's currently not really possible to provide custom configurations. It would also be hard to capture all environment variables used in the Graph API (specified in the CLI arguments), e.g. `HASH_LOG_FORMAT`. 
How do we want to pass `HASH_KRATOS_PG_USER` as `POSTGRES_USER`, setting it explicitly in the `env:` section?

File: packages/graph/Makefile.toml
Comment (TimDiekmann) [reply packages/graph/Makefile.toml:12]: Oh, sorry, GH  was misleading, I thought you were about the `touch-environment`.
We use `sleep-for-five-seconds` to spin-up the postgres instance inside of hash-graph

Comment (thehabbos007) [reply packages/graph/Makefile.toml:12]: Oh, no sorry the `[tasks.sleep-for-five-seconds]`, github adds too many lines above..

File: packages/graph/hash_graph/README.md
Comment (benwerner01) [inline packages/graph/hash_graph/README.md:16]: Not sure if I'm missing something, but why are we mentioning this script here? It was my understanding we would treat the `yarn external-services up` docker-compose configuration as the configuration for setting up the developer environment of the HASH workspace, not the developer environment of the HASH graph.

Also note that this script has to be run in the root of the repository, whereas the script it's replacing would've been run in this directory.

File: packages/graph/Makefile.toml
Comment (TimDiekmann) [reply packages/graph/Makefile.toml]: Was able to remove it: 2272c2ac4e437e0e6bff321ff6cc4011b67ea485

File: packages/hash/external-services/docker-compose.yml
Comment (thehabbos007) [reply packages/hash/external-services/docker-compose.yml:38]: Alright, perhaps they manually trap the signal :shrug:

File: packages/hash/external-services/docker-compose.yml
Comment (benwerner01) [reply packages/hash/external-services/docker-compose.yml:73]: > How do we want to be explicit here? 

Similar to how I've done it for the configuration of the `kratos` container below, so something like:

```suggestion
    env_file:
      - POSTGRES_USER=${HASH_GRAPH_PG_USER}
      - POSTGRES_PASSWORD=${HASH_GRAPH_PG_PASSWORD}
      - ...
```

This helps makes it explicit which environment variables the container is consuming from the `.env` file. If any are missing, docker-compose logs a warning that an environment variable is missing.

File: packages/hash/external-services/docker-compose.yml
Comment (thehabbos007) [reply packages/hash/external-services/docker-compose.yml:56]: It's more that I think the check will run forever and not just until it's in a healthy. From the docs its seems that it won't stop after the first healthy state check?

File: packages/graph/hash_graph/README.md
Comment (benwerner01) [reply packages/graph/hash_graph/README.md:16]: It was also my understanding that longterm we want to move the `yarn external-services` and related workspace scripts into `packages/hash`, at which point I don't think you could call it from this directory anymore.

File: packages/hash/external-services/docker-compose.yml
Comment (TimDiekmann) [reply packages/hash/external-services/docker-compose.yml:73]: The problem here is that we also define environment variables in the Graph and it would be really annoying if you cannot specify environment variables just because it's in the docker container because it was forgotten to add it to this list. Being explicit about this would be a big maintenance task.

File: packages/hash/external-services/docker-compose.yml
Comment (thehabbos007) [reply packages/hash/external-services/docker-compose.yml]: We concluded off-GitHub that we'll just keep it at 2 seconds for now. The healthcheck is cheap and shouldn't be too much of a problem. Keeping the timeout as-is though

**Review by thehabbos007 — DISMISSED** [review]
Looks good to me. We should probably wait for @benwerner01's approval as well :+1:

File: packages/hash/external-services/docker-compose.yml
Comment (TimDiekmann) [reply packages/hash/external-services/docker-compose.yml]: As discussed off-GH, we ermove the `env_file:` section and specify it manually. This is a bit annoying when adding more environment variables for `hash_graph`, but should be more robust with large `.env` files.

File: packages/graph/hash_graph/README.md
Comment (TimDiekmann) [reply packages/graph/hash_graph/README.md]: As discussed off-GH, we remove the entire section on running `graph` with docker. We expect a user for `graph` to use `cargo run`. For the HASH package, the external services are used.

File: packages/graph/deployment/graph/migrate_and_run.sh
Comment (benwerner01) [inline packages/graph/deployment/graph/migrate_and_run.sh:7]: I think we should use the root credentials to run the migrations on the database, incase we restrict permissions of the `HASH_GRAPH_USER` in a way that inhibits its ability to create/update tables/columns.
```suggestion
HASH_GRAPH_PG_MIGRATION_URL=postgres://${POSTGRES_USER}:${POSTGRES_PASSWORD}@${HASH_GRAPH_HOST}:${HASH_GRAPH_API_PORT}/${HASH_GRAPH_DATABASE} yarn graph:migrate up
```

Because the migrations are currently run in the `graph` container, that would mean we have to (temporarily) pass the root postgres credentials to that container (i.e `POSTGRES_USER` and `POSTGRES_PASSWORD`)

File: packages/hash/external-services/docker-compose.yml
Comment (benwerner01) [inline packages/hash/external-services/docker-compose.yml:82]: I think we should use the existing environment variables for these:
```suggestion
      HASH_GRAPH_HOST: ${POSTGRES_HOST}
      HASH_GRAPH_API_PORT: ${POSTGRES_PORT}
```

**Review by benwerner01 — COMMENTED** [review]
Thanks Tim, this is working for me locally and is looking great - I've spotted two more minor things which are probably worth addressing as part of this PR.

File: packages/hash/external-services/docker-compose.yml
Comment (TimDiekmann) [reply packages/hash/external-services/docker-compose.yml:82]: That's not possible, they are picked based on the docker environment. `"postgres"` refers to the other service and `"5432"` is a hard-coded port. `"${POSTGRES_PORT}"` is the port exposed from docker.

Comment (benwerner01) [reply packages/hash/external-services/docker-compose.yml:82]: Good point on the host having to be `"postgres"`, but isn't `"${POSTGRES_PORT}"` the port exposed from the docker container? i.e. any other container would have to use the value of `"${POSTGRES_PORT}"` to connect to it?

My understanding is that this isn't hardcoded, and is set [on line 56 of this file](https://github.com/hashintel/hash/blob/3c64a7c04cd1c55a1ee9304928f2dbb58f2a0a2c/packages/hash/external-services/docker-compose.yml#L56)

File: packages/graph/deployment/graph/migrate_and_run.sh
Comment (benwerner01) [reply packages/graph/deployment/graph/migrate_and_run.sh:7]: I'm wondering the same thing, it seems the `init-user-db.sh` script is not setting up permissions as intended

File: packages/hash/external-services/docker-compose.yml
Comment (TimDiekmann) [reply packages/hash/external-services/docker-compose.yml:82]: No, it's not set in 56. `POSTGRES_PORT` and `POSTGRES_HOST` is visible for the host network, it's forwarding the port from the docker network. However, the Graph is not in the host network but in the compose network and one has to specify the network for the postgres service (`"postgres"`). When specify a different `POSTGRES_PORT` and also setting this here, the migration script is failing, this needs to be postgres:5432. Anyway, I another weird error when using a different port: I cannot connect to the REST API for the Graph at all. Getting the OpenAPI specs is working.

File: packages/graph/deployment/graph/migrate_and_run.sh
Comment (TimDiekmann) [reply packages/graph/deployment/graph/migrate_and_run.sh]: Fixed in ae6fa88fbcf64bb62e3eaa6bec81849216fb2564

File: packages/graph/deployment/graph/Dockerfile
Comment (TimDiekmann) [reply packages/graph/deployment/graph/Dockerfile:26]: That's `HEALTHCHECK`, stupid bot 🤣

File: package.json
Comment (TimDiekmann) [inline package.json:33]: This took me about two hours because `docker-compose` is an alias to `docker compose` for macOS. On GH Actions, `docker-compose up --wait` will fail, but `docker compose up --wait` will not. 😒 😡

Comment (Alfred-Mountfield) [reply package.json:33]: God that is so frustrating

File: packages/graph/Makefile.toml
Comment (Alfred-Mountfield) [inline packages/graph/Makefile.toml:22]: Let's not do it in this PR but I think we should be extending the `tasks.yarn` task like we do for the `cargo` ones. We need to clean up this Makefile a bit, I've got some changes on a local branch to introduce the Makefile on another repo, can port them over.

File: packages/graph/Makefile.toml
Comment (Alfred-Mountfield) [inline packages/graph/Makefile.toml:53]: Do we want to be adding ` "${@}"` to all of these `args`

File: packages/graph/README.md
Comment (Alfred-Mountfield) [inline packages/graph/README.md:69]: Has this changed? I don't think it did require the graph to be running

File: packages/graph/Makefile.toml
Comment (TimDiekmann) [reply packages/graph/Makefile.toml:53]: Yes, that's probably a good idea. I'd love to be able to also pass additional parameters from other tasks but I don't think this is possible

File: packages/graph/deployment/graph/Dockerfile
Comment (Alfred-Mountfield) [reply packages/graph/deployment/graph/Dockerfile:26]: I mean it has commented on the `CMD ...` :D

File: packages/graph/README.md
Comment (TimDiekmann) [reply packages/graph/README.md:69]: Yes, it's required, it's needed that the binary is running in order to run `httpyac`.

I moved the README.md from `hash_graph` to the top level as this only contained one statement now.

File: packages/graph/deployment/graph/Dockerfile
Comment (TimDiekmann) [reply packages/graph/deployment/graph/Dockerfile:26]: I think it commented on both lines 🤷🏻‍♂️

File: packages/graph/hash_graph/lib/graph/src/store/postgres/pool.rs
Comment (Alfred-Mountfield) [inline packages/graph/hash_graph/lib/graph/src/store/postgres/pool.rs:49]: I can't see the struct but does this print passwords or anything?

Comment (TimDiekmann) [reply packages/graph/hash_graph/lib/graph/src/store/postgres/pool.rs:49]: No, password will be masked with `***`

File: packages/graph/README.md
Comment (Alfred-Mountfield) [reply packages/graph/README.md:69]: Before we had some code that started the binary and trapped it, it seems it was removed prior to this PR though. No worries

File: packages/hash/external-services/docker-compose.yml
Comment (benwerner01) [inline packages/hash/external-services/docker-compose.yml:66]: Could we rename this so that it's easier to distinguish between the containers running the graph migrations and the kratos migrations? If we want to be consistent with the kratos container name:
```suggestion
  graph-migrate:
```

File: packages/hash/external-services/docker-compose.yml
Comment (kachkaev) [reply packages/hash/external-services/docker-compose.yml:56]: We discussed this with @TimDiekmann in Slack and agreed that we could switch to a volume. A local folder is used to make sure that all temp files are contained in the `var` folder. We upload this folder as a CI artifact for debugging. A volume will change this, but we can think of a fix later if we do need to inspect the context of the volume.

File: packages/graph/Makefile.toml
Comment (TimDiekmann) [reply packages/graph/Makefile.toml]: Applied in 2b88cd0e39b010d1eb42fa3d6423a86c7f3a24c4

File: packages/hash/external-services/docker-compose.yml
Comment (TimDiekmann) [reply packages/hash/external-services/docker-compose.yml]: Applied in 12b73aedd3da04795e81d4a0a4047578c98cbff4

File: .env
Comment (benwerner01) [inline .env:15]: **Blocking:** This is currently being used by the workspace backend and the workspace integration tests to connect to the Graph API. Was this accidentally removed?

**Review by benwerner01 — CHANGES_REQUESTED** [review]
Thanks Tim, the dedicate `graph-migrate` container is a great addition 🎉

I think there was an accidental removal of a required environment variable, but if you bring that back everything is working locally for me as expected 👍

File: packages/hash/external-services/postgres/init-user-db.sh
Comment (TimDiekmann) [reply packages/hash/external-services/postgres/init-user-db.sh]: Good question!
By default everyone is free to create tables inside of the `public` schema. This revokes this right, so it's not possible anymore to create tables in the `public` schema from any `public` role. I changed the casing so it should be more clear now: cb249ef9151084ca719160f8b3538c176b70c666
See also [5.7.4. Schemas and Privileges](https://www.postgresql.org/docs/9.4/ddl-schemas.html#DDL-SCHEMAS-PRIV)

File: .env
Comment (TimDiekmann) [reply .env:15]: Yes, that was by accident (`POSTGRES_HOST` was on purpose as you shouldn't need to access `postgres` manually and the Dockerfile uses the network name instead.)
a2a18148040e7f67ccc53851bddbd052d715dd7c

Comment (benwerner01) [reply .env:15]: Great, thank you 👍

File: .env
Comment (benwerner01) [inline .env:27]: Nit: whitespace between the `.env` variables dedicated to specific services
```suggestion
HASH_GRAPH_PG_DATABASE=graph

HASH_GRAPH_API_HOST=localhost
HASH_GRAPH_API_PORT=4000
```

**Review by benwerner01 — APPROVED** [review]
Thank you Tim for ploughing through this, the setup looks great to me and is working for me locally 👍

**Review by benwerner01 — APPROVED** [review]

**Review by benwerner01 — APPROVED** [review]

File: packages/graph/Makefile.toml
Comment (nonparibus) [reply packages/graph/Makefile.toml:22]: @Alfred-Mountfield Did we capture this in a ticket to `@todo`? (just wanted to check)

**Pull request merged**
